client/src/entrypoints/admin/preview-panel.js

```diff
@@ -0,0 +1,236 @@
+    iframe.insertAdjacentElement('afterend', newIframe);
+
+    const handleLoad = () => {
+      // Copy all attributes from the old iframe to the new one,
+      // except src as that will cause the iframe to be reloaded
+      Array.from(iframe.attributes).forEach((key) => {
+        if (key.nodeName === 'src') return;
+        newIframe.setAttribute(key.nodeName, key.nodeValue);
+      });
+
+      // Restore scroll position
+      newIframe.contentWindow.scroll(
+        iframe.contentWindow.scrollX,
+        iframe.contentWindow.scrollY,
+      );
+
+      // Remove the old iframe and swap it with the new one
+      iframe.remove();
+      iframe = newIframe;
+
+      // Make the new iframe visible
+      newIframe.style = null;
+
+      // Ready for another update
+      finishUpdate();
+
+      // Remove the load event listener so it doesn't fire when switching modes
+      newIframe.removeEventListener('load', handleLoad);
+    };
+
+    newIframe.addEventListener('load', handleLoad);
+  };
+
+  const setPreviewData = () => {
+    hasPendingUpdate = true;
+    spinnerTimeout = setTimeout(
+      () => loadingSpinner.classList.remove('w-hidden'),
+      2000,
+    );
+
+    return fetch(previewUrl, {
+      method: 'POST',
+      body: new FormData(form),
+    })
+      .then((response) => response.json())
+      .then((data) => {
+        previewPanel.classList.toggle(
+          'preview-panel--has-errors',
+          !data.is_valid,
+        );
+        previewPanel.classList.toggle(
+          'preview-panel--unavailable',
+          !data.is_available,
+        );
+
+        if (data.is_valid) {
+          reloadIframe();
+        } else {
+          finishUpdate();
+        }
+
+        return data.is_valid;
+      })
+      .catch((error) => {
+        finishUpdate();
+        // Re-throw error so it can be handled by handlePreview
+        throw error;
+      });
+  };
+
+  const handlePreview = () =>
+    setPreviewData().catch(() => {
+      // eslint-disable-next-line no-alert
+      window.alert(gettext('Error while sending preview data.'));
+    });
+
+  const handlePreviewInNewTab = (event) => {
+    event.preventDefault();
+    const previewWindow = window.open('', previewUrl);
+    previewWindow.focus();
+
+    handlePreview().then((success) => {
+      if (success) {
+        const url = new URL(newTabButton.href);
+        previewWindow.document.location = url.toString();
+      } else {
+        window.focus();
+        previewWindow.close();
+      }
+    });
+  };
+
+  newTabButton.addEventListener('click', handlePreviewInNewTab);
+
+  if (previewPanel.dataset.autoUpdate === 'true') {
+    let oldData = new FormData(form);
+    const hasChanges = () => {
+      const newData = new FormData(form);
+      const oldPayload = new URLSearchParams(oldData).toString();
+      const newPayload = new URLSearchParams(newData).toString();
+
+      oldData = newData;
+      return oldPayload !== newPayload;
+    };
```

Comment:
I’m surprised this is so different from the existing dirty form check (I’d have expected us to reuse it exactly as-is but if there are reasons not to that’s fine). In particular:

- The existing check has logic to filter out comments. Feels like it’d be nice to do the same here?
- It has comparisons based on length rather than serialised field values (which I assume were done for performance reasons, not sure if warranted or not)
- It has a fair bit of logic to defer further checks when the form is dirty

For what it’s worth, I got measurements between 2ms and 8ms for `hasChanges` when testing this briefly, so this feels acceptable, but that was on a pretty small page.